mingpt/fake_lightning.py

```diff
@@ -0,0 +1,181 @@
+
+        def log(self, key, val):
+            pass
+
+class TrainResult(Result):
+    pass
+
+class LightningModule(nn.Module):
+    """ a very thin wrapper on module that returns the model's device similar to Lightning """
+
+    @property
+    def device(self):
+        return 'cuda' if next(self.parameters()).is_cuda else 'cpu'
+
+    @device.setter
+    def device(self, new_device):
+        raise RuntimeError("Cannot be set directly, is derived based on the module's parameters' location")
+
+class Callback:
+    pass
+
+class LightningDataModule:
+
+    def prepare_data(self):
+        pass
+
+    def setup(self, stage):
+        pass
+
+    def train_dataloader(self):
+        return None
+
+    def val_dataloader(self):
+        return None
+
+    def test_dataloader(self):
+        return None
+
+# -----------------------------------------------------------------------------
+"""
+Simple Trainer object; Boilerplate that could apply to any arbitrary neural network,
+so nothing here really has anything to do with GPT specifically. This is a
+very basic Trainer class that will only train the model on up to one GPU.
+"""
+
+class Trainer:
+
+    def __init__(self, max_epochs, gpus=0, gradient_clip_val=None, ckpt_path=None, callbacks=None,
+                 precision=32, **kwargs):
+        self.gpus = gpus
+        self.max_epochs = max_epochs
+        self.gradient_clip_val = gradient_clip_val
+        self.ckpt_path = ckpt_path
+        self.callbacks = [] if callbacks is None else callbacks
+        self.model = None
+
+        if self.gpus > 1:
+            logger.error("This simple Trainer does not support > 1 GPUs, will just use one.")
+
+        if precision != 32:
+            logger.error("This simple Trainer does not support non-fp32 precision, will use fp32")
+
+    def save_checkpoint(self):
+        # DataParallel wrappers keep raw model object in .module attribute
+        logger.info("saving model checkpoint to %s", self.ckpt_path)
+        torch.save(self.model.state_dict(), self.ckpt_path)
+
+    def load_checkpoint(self):
+        logger.info("loading the best model checkpoint from %s", self.ckpt_path)
+        state_dict = torch.load(self.ckpt_path)
+        self.model.load_state_dict(state_dict)
+
+    def fit(self, model, data_module):
+        self.model = model # bind model to the class here
+
+        # prepare the dataloaders for outputting batches
+        data_module.prepare_data()
+        data_module.setup('train')
+        train_loader = data_module.train_dataloader()
+        data_module.setup('val')
+        val_loader = data_module.val_dataloader()
+        data_module.setup('test')
+        test_loader = data_module.test_dataloader()
+
+        # ship model to gpu if possible
+        device = 'cpu'
+        if self.gpus > 0 and torch.cuda.is_available():
+            logger.info("found CUDA device, shipping model to GPU")
+            device = 'cuda'
+            self.model = self.model.to(device)
```

Comment:
if you are enabling gpus=x i would just do ```model.cuda(x)``` so people can place models on a gpu indexed by the PCI_BUS_ID (you might need this flag enabled though

```
export CUDA_DEVICE_ORDER=PCI_BUS_ID
```

Reply:
but for teaching purposes it may be overkill

Reply:
Thank you, yes it's cleaner. I like the more general .to syntax because we'll have many different XPUs etc, feels a bit more future proof